---
title: "pcadapt"
author: "Alyssa Green"
date: "`r Sys.Date()`"
output: html_document
---

Load Libraries
```{r}
library(RcppCNPy)
library(dplyr)
library(ggplot2)
library(BiocManager)
library(qvalue)
source("http://bit.ly/theme_pub") # By Dr. Rob Colautti
```

Load Data
```{r}
#Load in .npy file
Zscores=npyLoad("pcadapt_selection.pcadapt.zscores.npy")

#Load in .args file
args=read.csv("pcadapt_selection.args")

#Load in .cov file
cov=read.table("pcadapt_selection.cov")

#Load in .sites file
sites<-read.table("kittiwake.sites", row.names = NULL)
names(sites)<-c("chromosome","marker")

#Load in alt .sites file
alt_sites=read.table("pcadapt_selection.sites")

#Load Genes File
Genes=read.csv("Genes.csv")
```


Convert test statistics to p-values (following pcadapt_script.R)
```{r}
source("./pcadapt_script_PC.R")
PCp=read.table("PC.pcadapt.pval.txt")
```

Create QQplot to QC Test Statistics
```{r}
#Run qqchi function
source("./qqchi_func.R")
qqchi(PCp$V1)
```


Adjusting p values to q values following Caplins, S. (2021). Marine Genomics 2021. https://baylab.github.io/MarineGenomics/week-7--fst-and-outlier-analysis.html#finding-outliers-using-pcadapt
```{r}
qval <- qvalue(PCp)$qvalues
outliers <- which(qval<0.1)
length(outliers)
```

Aligning sites files
```{r}
sites$V1=alt_sites$V1
```

Remove elements of .sites values where V1=0
```{r}
sig=sites%>%filter(V1==1)
```

Create data frame combining sites and pvals
```{r}
Combo=data.frame(marker=sig$marker, chromosome=sig$chromosome, qval=qval$V1)
```

Select markers of interest
```{r}
to_keep = c()
gene_annotation = c()
scaffold=c()
q=c()
group=c()


  for (i in 1:length(Combo$chromosome)) {
  for (j in 1:length(Genes$Code)){
      if (Genes$Code[j] == Combo$chromosome[i]){
        if(between(Combo$marker[i], Genes$Min[j], Genes$Max[j])){
            to_keep = c(to_keep,i)
            gene_annotation = c(gene_annotation,Genes$Gene[j])
            scaffold=c(scaffold, Combo$chromosome[i])
            q=c(q, Combo$qval[i])
            group=c(group, Genes$Group[j])
            
              }
         }
    }
}

```

Create data frame with information for genes on interest
```{r}
Data=data.frame(gene_annotation, scaffold, marker=to_keep, q, group)
head(Data)

```

Number of significant hits, and filter non-significant results from data set
```{r}
outliers <- which(Data$q<1.07*10^{-9})
length(outliers)
Outliers=Data%>%filter(q<1.07*10^{-9})
```

Find number of outliers for each gene and order from most to least outliers
```{r}
OutGene=Outliers%>%group_by(gene_annotation)%>%summarise(total_count=n(),
            .groups = 'drop')
OutGeneOrder=OutGene%>%arrange(desc(total_count))

head(OutGeneOrder)
```

Check number of genes with significant results
```{r}
no_duplicates=distinct(Outliers, gene_annotation, .keep_all = TRUE)
dim(no_duplicates)
```

Check name and number of genes in each gene group
```{r}
rp=filter(no_duplicates, group == "RNA Pol")
dp=filter(no_duplicates, group == "DNA Pol")
thy=filter(no_duplicates, group == "Thyroid")
glu=filter(no_duplicates, group == "Glucocorticoid")
tel=filter(no_duplicates, group == "Telomeres")
cB=filter(no_duplicates, group == "CytB")
mel=filter(no_duplicates, group == "Melanocortins")

nrow(rp)
nrow(dp)
nrow(thy)
nrow(glu)
nrow(tel)
nrow(cB)
nrow(mel)
```


Manhattan Plot
```{r, fig.width=15, fig.height=3}
ggplot(aes(x=gene_annotation, y=-log10(q), colour=group), data=Data)+geom_point()+theme_pub()+geom_hline(yintercept=-log10(1.07*10^(-9)), colour="black")+xlab("Gene")+ylab("-log10(q-value)")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=2.5))+theme(plot.title = element_text(hjust = 0))
```
